servers/src/main/java/tachyon/worker/block/BlockDataManager.java

```diff
@@ -0,0 +1,450 @@
+    String dstPath = CommonUtils.concatPath(ufsDataFolder, fileId);
+    try {
+      if (!mUfs.rename(srcPath, dstPath)) {
+        throw new FailedToCheckpointException("Failed to rename " + srcPath + " to " + dstPath);
+      }
+    } catch (IOException e) {
+      throw new FailedToCheckpointException("Failed to rename " + srcPath + " to " + dstPath);
+    }
+    long fileSize;
+    try {
+      fileSize = mUfs.getFileSize(dstPath);
+    } catch (IOException e) {
+      throw new FailedToCheckpointException("Failed to getFileSize " + dstPath);
+    }
+    mMasterClient.addCheckpoint(mWorkerId, fileId, fileSize, dstPath);
+  }
+
+  /**
+   * Cleans up after users, to prevent zombie users. This method is called periodically.
+   */
+  public void cleanupUsers() throws IOException {
+    for (long user : mUsers.getTimedOutUsers()) {
+      mUsers.removeUser(user);
+      mBlockStore.cleanupUser(user);
+    }
+  }
+
+  /**
+   * Commits a block to Tachyon managed space. The block must be temporary. The block is
+   * persisted after {@link BlockStore#commitBlock(long, long)}. The block will not be accessible
+   * until {@link MasterClient#worker_cacheBlock(long, long, long, long, long)}
+   * succeeds
+   *
+   * @param userId The id of the client
+   * @param blockId The id of the block to commit
+   * @return true if successful, false otherwise
+   * @throws IOException if the block to commit does not exist
+   */
+  public void commitBlock(long userId, long blockId) throws IOException {
+    mBlockStore.commitBlock(userId, blockId);
+
+    // TODO: Reconsider how to do this without heavy locking
+    // Block successfully committed, update master with new block metadata
+    Long lockId = mBlockStore.lockBlock(userId, blockId);
+    try {
+      BlockMeta meta = mBlockStore.getBlockMeta(userId, blockId, lockId);
+      BlockStoreLocation loc = meta.getBlockLocation();
+      Long storageDirId = loc.getStorageDirId();
+      Long length = meta.getBlockSize();
+      BlockStoreMeta storeMeta = mBlockStore.getBlockStoreMeta();
+      Long bytesUsedOnTier = storeMeta.getUsedBytesOnTiers().get(loc.tierLevel());
+      mMasterClient
+          .worker_cacheBlock(mWorkerId, bytesUsedOnTier, storageDirId, blockId, length);
+    } catch (TException te) {
+      throw new IOException("Failed to commit block to master.", te);
+    } finally {
+      mBlockStore.unlockBlock(userId, blockId);
+    }
+  }
+
+  /**
+   * Creates a block in Tachyon managed space. The block will be temporary until it is committed.
+   *
+   * @param userId The id of the client
+   * @param blockId The id of the block to create
+   * @param tierAlias The alias of the tier to place the new block in, -1 for any tier
+   * @param initialBytes The initial amount of bytes to be allocated
+   * @return A string representing the path to the local file
+   * @throws IOException if the block already exists
```

Comment:
the `IOException` can mean more exception situations than this, but it seems like we should add more explicit Exceptions in following PR.